### PR TITLE
[net-diag] introduce `Server` and `Client` classes

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces `Server` and `Client` classes breaking the
`NetworkDiagnostic` module into two components. The `Server` responds
to queries and requests (handling Net Diag TMF commands), while
`Client` issues queries/requests and processes responses. The
`Server` is available under both FTD and MTD (which follows the
requirement of Thread spec). The client is enabled using newly added
`OPENTHREAD_CONFIG_TMF_NETDIAG_CLIENT_ENABLE` config option which is
also available as `OT_NETDIAG_CLIENT` CMake option or the autoconf
build switch `NETDIAG_CLIENT`. The client functionality is by default
enabled on Border Routers (tied to `CONFIG_BORDER_ROUTING_ENABLE`
config).

The previous `TMF_NETWORK_DIAG_MTD_ENABLE` config is now removed
along with its CMake `OT_MTD_NETDIAG` and autoconf switch. This
commit adds checks to trigger build error if the previous config
and/or its related CMake option are used.

----

I expect this change will create a conflict for some of other open PRs related to netdiag, but feel better to have this done earlier (we can address the conflicts in other PRs). :)

### CODE REVIEW

**Comment by jwhui** [comment]
Thanks for this enhancement!

> The Client is always included on FTD.

What is the resource overhead for the Client? If significant, we could consider enabling by default only on border routers. I'm wondering if the Client is useful on something like a light bulb or smart plug.

**Comment by abtink** [comment]
> What is the resource overhead for the Client? If significant, we could consider enabling by default only on border routers. I'm wondering if the Client is useful on something like a light bulb or smart plug.

Good point. I will check the code size impact.

File: etc/gn/openthread.gni [hunk 1]
```diff
@@ -174,8 +174,8 @@ if (openthread_enable_core_config_args) {
     # Enable MLE long routes extension (experimental, breaks Thread conformance]
     openthread_config_mle_long_routes_enable = false
 
-    # Enable TMF network diagnostics on MTDs
-    openthread_config_tmf_network_diag_mtd_enable = false
+    # Enable TMF network diagnostics clinet
```
Comment (jwhui) [inline etc/gn/openthread.gni:177]:
```suggestion
    # Enable TMF network diagnostics client
```

File: src/core/thread/tmf.cpp [hunk 2]
```diff
@@ -164,11 +164,11 @@ bool Agent::HandleResource(const char *aUriPath, Message &aMessage, const Ip6::M
         Case(kUriAnycastLocate, AnycastLocator);
 #endif
 
-#if OPENTHREAD_FTD || OPENTHREAD_CONFIG_TMF_NETWORK_DIAG_MTD_ENABLE
-        Case(kUriDiagnosticGetRequest, NetworkDiagnostic::NetworkDiagnostic);
-        Case(kUriDiagnosticGetQuery, NetworkDiagnostic::NetworkDiagnostic);
-        Case(kUriDiagnosticGetAnswer, NetworkDiagnostic::NetworkDiagnostic);
-        Case(kUriDiagnosticReset, NetworkDiagnostic::NetworkDiagnostic);
+#if OPENTHREAD_CONFIG_TMF_NETDIAG_CLIENT_ENABLE
+        Case(kUriDiagnosticGetRequest, NetworkDiagnostic::Server);
```
Comment (jwhui) [inline src/core/thread/tmf.cpp:168]: These are server-side handlers. Why are they guarded by `CLIENT_ENABLE`?

Comment (abtink) [reply src/core/thread/tmf.cpp:168]: Thanks. That's not correct. I missed this. Will fix in new push.

**Review by jwhui — APPROVED** [review]

**Pull request merged**